common/edl/CMakeLists.txt

```diff
@@ -0,0 +1,46 @@
+# Copyright (c) Open Enclave SDK contributors.
+# Licensed under the MIT License.
+
+add_library(oeedl_includes INTERFACE)
```

Comment:
Setting up a separate edl folder under common for just these types seems like it adds another way that EDL is handled differently in the CMake project and I would really like to minimize that where possible. A couple of observations:

* The common folder has EDL files at the top level and also in the sgx folder, so having an `edl` folder that doesn't contain all the EDLs seems misleading. There's a reason for this folder that's not reflected in the name, nor explained elsewhere.

   * If you're going to keep this folder, it might be useful to at least call out in a README.md for the folder that these type definitions are special because they are included by the dev-authored, public-facing host.h & enclave.h headers. (Otherwise they could just be consumed like other regular args.h from the EDL build process).

* It looks like the only reason this is in a separate directory to to have a separate CMake target for this file ... why?

   * For example, an alternative might be that these data types are declared in tee.edl, which is already built, and then the extra `oeedl_includes` library target and custom command to copy and rename the resulting `tee_args.h` to includes could be done in either the host or enclave builds.

Reply:
The reason to have a separated edl with only data structures is trying make the published header as clean as possible. For example, the `tee_args.h` will contain several ECALL/OCALL marshalling structs, which I tried to avoid in the published header.

The end goal is to publishing all the data structures used by ECALLs/OCALLs in a similar way, which will expands the list. We can definitely move all the EDL files under `common` into the `common/edl` folder given that the list of EDL is increasing because of recent refactoring work.

Reply:
I'm in favor of moving all the EDL files under this folder.

Reply:
Related question to this decision.
If we consider allowing a developer to import any of internal edl files, then we need to expose those edl files as part of the installation package.

Take an example from this PR. If a developer wants to use `oe_asymmetric_key_params_t` as part of his ECALLs/OCALLs, he then needs to find the edl file somewhere and import.

Something like:

`from "{installdir}/include/openenclave/edl/asym_keys.edl" import *`

or

`from "{installdir}/edl/asym_keys.edl" import *`

or 

`from "{installdir}/common/edl/asym_keys.edl" import *`


@jhand2 What's the plan to opt-in OCALLs?

Reply:
@jhand2 
> @CodeMonkeyLeet was there a desire to treat EDL the same as C files and not split it out in the directory structure? If so I would also be fine with keeping the directory structure as-is.

I think EDL files are already being treated in a special way regardless of where they are located based on the existing host/CMakeLists.txt and enclave/CMakeLists.txt, my concern is that we're introducing yet another special way to build EDL files and consume them in this special subdirectory. 

To be clear, I don't feel strongly that we need to resolve it in this PR, but I think it points to additional work we need to more broadly with regards to how OE consumes headers generated by EDL to enable deep-copy scenarios. Here's a couple of other pivots:

* I had a suggestion that these data structures should just be a part of tee.edl and then tee_args.h published directly to the bits folder. There could be concern that not all of the args in tee_args.h need to be published (e.g. `oe_log_ocall_args_t`; note that function headers are not published in *_args.h only the argument types). I think the way to address that should actually be to split the functions and its data types out into their own EDL, because it's really the functions themselves that create this dependency. 

* At a higher level, some of these methods are largely exposed through EDL just to provide convenience methods on the host side to invoke enclave functions, and it's becoming more clear that maybe that just complicates the API surface if the use cases are not actually in use (e.g. `oe_get_report` from the host side is such a candidate, given its different semantics from the enclave variant)

Reply:
@jhand2 Do you expect the list of such data structure to grow (i.e., the one used by internal ECALLs needs to be published)? If this is a special case, then I agree not new creating this new cmake target for that.

@CodeMonkeyLeet Are you fine with the published headers containing ECALL/OCALL marshalling structs?

Reply:
> @jhand2
> 
> > @CodeMonkeyLeet was there a desire to treat EDL the same as C files and not split it out in the directory structure? If so I would also be fine with keeping the directory structure as-is.
> 
> I think EDL files are already being treated in a special way regardless of where they are located based on the existing host/CMakeLists.txt and enclave/CMakeLists.txt, my concern is that we're introducing yet another special way to build EDL files and consume them in this special subdirectory.
> 
> To be clear, I don't feel strongly that we need to resolve it in this PR, but I think it points to additional work we need to more broadly with regards to how OE consumes headers generated by EDL to enable deep-copy scenarios. Here's a couple of other pivots:
> 
> * I had a suggestion that these data structures should just be a part of tee.edl and then tee_args.h published directly to the bits folder. There could be concern that not all of the args in tee_args.h need to be published (e.g. `oe_log_ocall_args_t`; note that function headers are not published in *_args.h only the argument types). I think the way to address that should actually be to split the functions and its data types out into their own EDL, because it's really the functions themselves that create this dependency.

This is the direction I thought initially, which cleanly addresses my concern. I'd like to go for it.

> * At a higher level, some of these methods are largely exposed through EDL just to provide convenience methods on the host side to invoke enclave functions, and it's becoming more clear that maybe that just complicates the API surface if the use cases are not actually in use (e.g. `oe_get_report` from the host side is such a candidate, given its different semantics from the enclave variant)

Reply:
@mingweishih I'm okay with it as a tradeoff for a more consistent build flow. I agree with @jhand2 that having the data structures declared separately is cleaner in terms of public includes, and I will defer to him if he feels strongly about it. 

Regarding the list of data structures, I suspect it will continue to grow for any function that needs to marshal an OE data structure across the boundary, which may not be unique to this function once we start pushing oecore functions out of manual handling and into EDL.

Reply:
@mingweishih I would prefer that structs which will be built into headers and published as part of the public API surface be kept in separate EDL files. For internal structs I think it's fine to keep them in the same EDL file as the functions that use them.